File: agentic_os/mission/compiler.py
"""Execution Compiler — the PHYSICAL layer (logical plan -> physical plan). NO LLM.

Binds an ExecutionIntent to concrete capabilities (registry + cost model), orders them into an
ExecutionGraph, injects human/permission/guardrail decisions, assigns idempotency keys, and
validates every node's permissions against the mission's grants (FAIL CLOSED). Because it is
deterministic it is cheap, replayable and unit-testable — the whole point of the split. If the
model ever needs to run here, the layering has leaked: push the decision up into the intent.
"""
from __future__ import annotations

import hashlib
import re

from .context import BIND_CAPABILITY, ContextIntent, ContextRuntime
from .types import (
    Mission, ExecutionIntent, ExecutionPlan, ExecutionGraph, Node, NodeCost, CapabilitySpec,
)


def _node_id(mission_id: str, revision: int, outcome: str) -> str:
    """Deterministic node id — so a graph RECOMPILED after a restart correlates with the
    event-sourced execution state (which is keyed by node id). Durability depends on this."""
    slug = re.sub(r"[^a-z0-9]+", "_", outcome.lower()).strip("_")[:32]
    return f"nd_{mission_id.split('_')[-1][:8]}_{revision}_{slug}"


class CompileError(Exception):
    """Raised when an intent cannot be compiled — no capability, or a permission is not granted."""


def _granted(mission: Mission) -> set[str]:
    return set(mission.policy_refs or [])


def _permitted(cap: CapabilitySpec, granted: set[str]) -> bool:
    if "*" in granted:
        return True
    return all(p in granted for p in cap.permissions)


# Capability binding no longer lives here — it is a CONTEXT decision. The compiler states the need and
# the Context Runtime resolves it (which provider, over provides vs embedding discovery). See context.py.


def _idem(mission_id: str, revision: int, outcome: str) -> str:
    return hashlib.sha256(f"{mission_id}:{revision}:{outcome}".encode()).hexdigest()[:16]


def compile_intent(
    mission: Mission,
    intent: ExecutionIntent,
    context: ContextRuntime,
    revision: int = 1,
    reason: str = "initial",
    prefer: dict[str, str] | None = None,
) -> ExecutionPlan:
    if not hasattr(context, "resolve"):
        # tolerate a bare registry (tests / legacy callers) — wrap it in the default Context Runtime,
        # so binding still resolves through the CR door rather than the compiler reaching the registry.
        from .context import LocalContextRuntime
        context = LocalContextRuntime(context)
    granted = _granted(mission)
    prefer = prefer or {}
    graph = ExecutionGraph()
    outcome_to_node: dict[str, str] = {}

    # First pass: bind every step to a capability + node. Binding is a CONTEXT decision — the compiler
    # states the need; the Context Runtime resolves which capability satisfies it.
    step_nodes: list[tuple] = []
    for step in intent.steps:
        cap = context.resolve(ContextIntent(
            kind=BIND_CAPABILITY, need=step.need, outcome=step.outcome,
            prefer=prefer.get(step.outcome), mission=mission)).value
        if cap is None:
            raise CompileError(f"no capability provides outcome '{step.outcome}' (need: {step.need})")
        if not _permitted(cap, granted):
            missing = [p for p in cap.permissions if p not in granted]
            raise CompileError(f"permission denied for '{cap.name}': missing grant(s) {missing}")
        # a node needs human approval if the capability demands it, or a constraint asks for review
        human = cap.approval_required or any("human" in c.lower() or "review" in c.lower()
                                             for c in step.constraints)
        node = Node(
            capability=cap.name, operator=cap.operator, produces=step.outcome,
            approval_required=human, side_effecting=cap.side_effecting, undo=cap.undo,
            idempotency_key=_idem(mission.id, revision, step.outcome),
            cost=NodeCost(usd=cap.cost.usd, latency_ms=cap.cost.latency_ms,
                          tokens=cap.cost.tokens, human_minutes=cap.cost.human_minutes),
            intent_step_id=step.id,
        )
        node.id = _node_id(mission.id, revision, step.outcome)   # deterministic (restart-safe)
        graph.nodes.append(node)
        outcome_to_node[step.outcome] = node.id
        step_nodes.append((step, node))

    # Second pass: wire data dependencies from inputs_from (outcome -> node id).
    for step, node in step_nodes:
        for upstream_outcome in step.inputs_from:
            up = outcome_to_node.get(upstream_outcome)
            if up:
                node.depends_on.append(up)
            # inputs the node will read from world state at run time
            node.inputs[upstream_outcome] = {"$from_world": upstream_outcome}

    _assert_acyclic(graph)
    return ExecutionPlan(mission_id=mission.id, intent_id=intent.id, graph=graph,
                         revision=revision, reason=reason)


def _assert_acyclic(graph: ExecutionGraph) -> None:
    indeg = {n.id: 0 for n in graph.nodes}
    for n in graph.nodes:
        for _ in n.depends_on:
            indeg[n.id] += 1
    ready = [nid for nid, d in indeg.items() if d == 0]
    seen = 0
    adj: dict[str, list[str]] = {n.id: [] for n in graph.nodes}
    for n in graph.nodes:
        for dep in n.depends_on:
            adj[dep].append(n.id)
    while ready:
        cur = ready.pop()
        seen += 1
        for nxt in adj[cur]:
            indeg[nxt] -= 1
            if indeg[nxt] == 0:
                ready.append(nxt)
    if seen != len(graph.nodes):
        raise CompileError("execution graph has a cycle")

File: go/mission/compiler.go
package mission

// Execution Compiler — the PHYSICAL layer (logical plan → physical plan). NO LLM. Binds an
// ExecutionIntent to concrete capabilities (registry + cost model), orders them into a graph,
// injects human/permission decisions, assigns idempotency keys, and validates every node's
// permissions against the mission's grants (FAIL CLOSED). Go port of compiler.py.

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"
)

// CompileError is raised when an intent can't be compiled (no capability, or a permission denied).
type CompileError struct{ Msg string }

func (e *CompileError) Error() string { return e.Msg }

// minBindScore: a discovery match below this similarity is NOT a real provider — binding an
// unrelated (but permitted) capability to a required outcome would be a silent correctness bug.
const minBindScore = 0.2

// nodeID is a deterministic node id so a graph RECOMPILED after a restart correlates with the
// event-sourced execution state (keyed by node id). Durability depends on this.
func nodeID(missionID string, revision int, outcome string) string {
	parts := strings.Split(missionID, "_")
	last := parts[len(parts)-1]
	if len(last) > 8 {
		last = last[:8]
	}
	return fmt.Sprintf("nd_%s_%d_%s", last, revision, slugify(outcome, 32))
}

func idemKey(missionID string, revision int, outcome string) string {
	h := sha256.Sum256([]byte(fmt.Sprintf("%s:%d:%s", missionID, revision, outcome)))
	return hex.EncodeToString(h[:])[:16]
}

func permittedCap(cap *CapabilitySpec, granted map[string]bool) bool {
	if granted["*"] {
		return true
	}
	for _, p := range cap.Permissions {
		if !granted[p] {
			return false
		}
	}
	return true
}

// bindCapability picks the best capability that PROVIDES the outcome, else a STRONG semantic match
// to the need. prefer (a capability name) forces a specific provider if it provides the outcome.
// Binding is a Context Runtime resolution: the compiler states the need and the runtime decides
// (mirrors compiler.py, which calls context.resolve(BIND_CAPABILITY)).
func bindCapability(stepNeed, outcome string, registry Registry, prefer string) *CapabilitySpec {
	b := NewLocalContextRuntime(registry).Resolve(ContextIntent{
		Kind: BindCapability, Need: stepNeed, Outcome: outcome, Prefer: prefer})
	cap, _ := b.Value.(*CapabilitySpec)
	return cap
}

// CompileIntent binds an intent to a physical ExecutionPlan (fail-closed on permissions).
func CompileIntent(mission *Mission, intent ExecutionIntent, registry Registry,
	revision int, reason string, prefer map[string]string) (*ExecutionPlan, error) {

	granted := map[string]bool{}
	for _, g := range mission.PolicyRefs {
		granted[g] = true
	}
	if prefer == nil {
		prefer = map[string]string{}
	}
	graph := NewExecutionGraph()
	outcomeToNode := map[string]string{}
	type sn struct {
		step IntentStep
		idx  int
	}
	var stepNodes []sn

	for _, step := range intent.Steps {
		cap := bindCapability(step.Need, step.Outcome, registry, prefer[step.Outcome])
		if cap == nil {
			return nil, &CompileError{Msg: fmt.Sprintf("no capability provides outcome %q (need: %s)", step.Outcome, step.Need)}
		}
		if !permittedCap(cap, granted) {
			var missing []string
			for _, p := range cap.Permissions {
				if !granted[p] {
					missing = append(missing, p)
				}
			}
			return nil, &CompileError{Msg: fmt.Sprintf("permission denied for %q: missing grant(s) %v", cap.Name, missing)}
		}
		human := cap.ApprovalRequired
		for _, c := range step.Constraints {
			lc := strings.ToLower(c)
			if strings.Contains(lc, "human") || strings.Contains(lc, "review") {
				human = true
				break
			}
		}
		node := NewNode(cap.Name, cap.Operator)
		node.Produces = step.Outcome
		node.ApprovalRequired = human
		node.SideEffecting = cap.SideEffecting
		node.Undo = cap.Undo
		node.IdempotencyKey = idemKey(mission.ID, revision, step.Outcome)
		node.Cost = cap.Cost
		// carry the capability's declared concurrency safety semantics onto the node (additive)
		node.ConcurrencyMode = cap.ConcurrencyMode
		node.ConcurrencyKey = cap.ConcurrencyKey
		node.ResourceKeys = append([]string(nil), cap.ResourceKeys...)
		node.MaxParallelism = cap.MaxParallelism
		node.IntentStepID = step.ID
		node.Inputs = map[string]any{}
		node.ID = nodeID(mission.ID, revision, step.Outcome) // deterministic (restart-safe)

		graph.Nodes = append(graph.Nodes, node)
		outcomeToNode[step.Outcome] = node.ID
		stepNodes = append(stepNodes, sn{step, len(graph.Nodes) - 1})
	}

	for _, s := range stepNodes {
		for _, up := range s.step.InputsFrom {
			if upID, ok := outcomeToNode[up]; ok {
				graph.Nodes[s.idx].DependsOn = append(graph.Nodes[s.idx].DependsOn, upID)
			}
			graph.Nodes[s.idx].Inputs[up] = map[string]any{"$from_world": up}
		}
	}

	if err := assertAcyclic(graph); err != nil {
		return nil, err
	}
	plan := NewExecutionPlan(mission.ID, intent.ID, graph)
	plan.Revision = revision
	plan.Reason = reason
	return plan, nil
}

// assertAcyclic is Kahn's algorithm — errors if the graph has a cycle.
func assertAcyclic(graph *ExecutionGraph) error {
	indeg := map[string]int{}
	adj := map[string][]string{}
	for i := range graph.Nodes {
		indeg[graph.Nodes[i].ID] = 0
	}
	for i := range graph.Nodes {
		n := &graph.Nodes[i]
		for _, dep := range n.DependsOn {
			indeg[n.ID]++
			adj[dep] = append(adj[dep], n.ID)
		}
	}
	var ready []string
	for nid, d := range indeg {
		if d == 0 {
			ready = append(ready, nid)
		}
	}
	seen := 0
	for len(ready) > 0 {
		cur := ready[len(ready)-1]
		ready = ready[:len(ready)-1]
		seen++
		for _, nxt := range adj[cur] {
			indeg[nxt]--
			if indeg[nxt] == 0 {
				ready = append(ready, nxt)
			}
		}
	}
	if seen != len(graph.Nodes) {
		return &CompileError{Msg: "execution graph has a cycle"}
	}
	return nil
}
